crates/router/src/core/payment_methods/surcharge_decision_configs.rs

```diff
@@ -424,12 +383,21 @@ pub async fn refresh_surcharge_algorithm_cache(
         .parse_struct("Program")
         .change_context(ConfigError::DslParsingError)
         .attach_printable("Error parsing routing algorithm from configs")?;
-    let value_to_cache = VirInterpreterBackendCacheWrapper::try_from(record)?;
-    CONF_CACHE
-        .save(key, value_to_cache, timestamp)
-        .change_context(ConfigError::DslCachePoisoned)
-        .attach_printable("Error saving DSL to cache")?;
-    Ok(())
+    let value_to_cache = || async {
+        VirInterpreterBackendCacheWrapper::try_from(record)
+            .change_context(errors::StorageError::ValueNotFound("Program".to_string()))
+            .attach_printable("Error initializing DSL interpreter backend")
+    };
```

Comment:
You can move the DB call to fetch the config (and thus parsing the record) within the closure, to avoid unnecessary DB calls.